JBQA-12612 - Test JDG70.HadoopIOFormat

andyuk1986 opened the pull request:
Added Jacoco plugin for calculating code coverage; 
Added more tests.

gustavonalle commented:
Regarding the changes on ```InputOutputFormatTest```, it is taking 7min to run rather than 1min on master.
Apparently the Infinispan Server is being stopped/started multiple times (before/after each test?), when on master it does only once for the whole suite.

andyuk1986 commented:
Ooops, sorry as I have changed the arquillian.xml configuration so that it is possible to start/stop servers manually, I have changed the InputOutpuFormatTest correspondingly and overlooked that. 

In JUnit the only option for starting/stopping before / after all tests are executed is @BeforeClass and @AfterClass annotations, but they are static, so for my case they were unusable. I didn't want to add TestNG dependency and change everything to it. So I have implemented a workaround - hope it is ok for you. If not, then lets discuss how I can handle this.

gustavonalle commented:
@andyuk1986 master had some changes regarding memory usage and speed in the tests (the whole suite now takes 30s in my env), and also a configuration change was introduced to provide better failover. 
Could you please rebase?

andyuk1986 commented:
@gustavonalle , I have rebased my changes. Also have updated the memory parameter (as with the set one the FailoverTests were giving outofmemoryexception). If the set parameter is not ok, then let me know - I'll change it back. 

The issue with failovers is fixed. But no the ISPN-6087 appears constantly for all failover cases. I.e. the count of given word is greater than expected. Is it possible that a duplicate processing is performed on some records after one of the servers stopped?

gustavonalle commented:
>  ISPN-6087 appears constantly for all failover cases

I have a feeling it's due to speculative execution in Hadoop. Could you try disabling it in the job?

andyuk1986 commented:
@gustavonalle what do you mean by speculative execution?

andyuk1986 commented:
@gustavonalle I have disabled the speculative execution, but the failure is still there.

gustavonalle commented:
@andyuk1986 Thanks, I will investigate.

gustavonalle commented on core/src/test/java/org/infinispan/hadoop/FailoverHandlingTest.java:
```diff
@@ -0,0 +1,300 @@
+public class FailoverHandlingTest {
+
+    private static final String INPUT_CACHE_NAME = "default";
+    private static final String OUTPUT_CACHE_NAME = "namedCache";
```
This cache was removed from Infinispan 8.2.0.CR1 so we need to use "memcachedCache" for now

gustavonalle commented:
Hi @andyuk1986, sorry for the delay, could you rebase? Tests should be passing!

andyuk1986 commented:
Hi @gustavonalle , I have changed the cache name to "memcachedCache" and rebased. The tests are passing. Thanks a lot for the fix :)

gustavonalle commented:
Sometimes I get exceptions during clear, created https://issues.jboss.org/browse/ISPN-6304
Still investigating the other failures on CI, them do not seem to happen locally...

gustavonalle commented:
Opened  https://issues.jboss.org/browse/ISPN-6459 and https://issues.jboss.org/browse/ISPN-6304 to fix the errors appearing on CI

gustavonalle commented:
ISPN-6304 and ISPN-6459 have both been fixed and the intermittent failures mentioned above are not happening anymore.

But I am picking another intermittent failure:
```
testReadFromInfinispanSaveToInfinispan_StopOutputServer(org.infinispan.hadoop.FailoverHandlingTest)  Time elapsed: 56.027 sec  <<< ERROR!
java.lang.NullPointerException: null
        at org.infinispan.hadoop.FailoverHandlingTest.executeTest(FailoverHandlingTest.java:285)
        at org.infinispan.hadoop.FailoverHandlingTest.testReadFromInfinispanSaveToInfinispan_StopOutputServer(FailoverHandlingTest.java:227)
```

It is not easy to reproduce, it requires pinning the test to a single CPU and running in a loop:
```
loop 10 tasket -c 0 mvn clean install
```
(loop is defined as  ```function loop { seq 1 $1 | { shift; xargs -I -- "$@";  } }```)

I will investigate it.

gustavonalle commented:
Integrated, thanks @andyuk1986 !
I created [IHDP-10](https://issues.jboss.org/browse/IHDP-10) to investigate the intermittent NPE.

andyuk1986 commented:
Cooool!!! Thanks a lot @gustavonalle  :))))
